ngrinder-core/src/main/java/net/grinder/SingleConsole.java

```diff
@@ -86,6 +92,9 @@
 
 	private final Condition eventSyncCondition = new Condition();
 	private ProcessReports[] processReports;
+
+	// It contains cached distribution files md5 checksum from each agents.
+	private List<Set<String>> agentCachedDistFilesMd5List = synchronizedList(new ArrayList<>());

```

Comment:
How about use Concurrent Collections rather than Synchronized Collections?
https://javarevisited.blogspot.com/2016/05/what-is-difference-between-synchronized.html

Reply:
When to use SynchronizedList?
1. Since in CopyOnWriteArrayList for every update/modify operations, a new separate cloned copy is created and there is overhead on JVM to allocate memory & merging cloned copy with original copy. Thus, in this case SynchronizedList is better option.
2. When size of Arraylist is large.

When to use CopyOnWriteArrayList?

1. The CopyOnWriteArrayList provides reading without a lock, which means a much better performance if there are more reader threads and write is happening quite low.
2. When the size of Arraylist is small.

In our case it is better to use CopyOnWriteArrayList because the list size is small and only use read operation after the first add to the list.